Jellyfin.Server.Implementations/UserData/UserDataManager.cs

```diff
@@ -0,0 +1,262 @@
+using MediaBrowser.Controller.Configuration;
+using MediaBrowser.Controller.Dto;
+using MediaBrowser.Controller.Entities;
+using MediaBrowser.Controller.Library;
+using MediaBrowser.Model.Dto;
+using MediaBrowser.Model.Entities;
+using Microsoft.EntityFrameworkCore;
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+
+namespace Jellyfin.Server.Implementations.UserData;
+
+/// <summary>
+/// Manages the storage and retrieval of <see cref="UserItemData"/> instances.
+/// </summary>
+public class UserDataManager : IUserDataManager
+{
+    private readonly IServerConfigurationManager _config;
+    private readonly IDbContextFactory<LibraryDbContext> _provider;
+    private readonly IUserManager _userManager;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="UserDataManager"/> class.
+    /// </summary>
+    /// <param name="config">The <see cref="IServerConfigurationManager"/> to use in the <see cref="UserDataManager"/> instance.</param>
+    /// <param name="provider">The Jellyfin database provider.</param>
+    /// <param name="userManager">The Jellyfin user manager.</param>
+    public UserDataManager(
+        IServerConfigurationManager config,
+        IDbContextFactory<LibraryDbContext> provider,
+        IUserManager userManager)
+    {
+        _config = config;
+        _provider = provider;
+        _userManager = userManager;
+    }
+
+    /// <inheritdoc />
+    public event EventHandler<UserDataSaveEventArgs>? UserDataSaved;
+
+    /// <summary>
+    /// Save UserData.
+    /// </summary>
+    /// <param name="user">The User object.</param>
+    /// <param name="item">The BaseItem which relates.</param>
+    /// <param name="userData">The info against the base item.</param>
+    /// <param name="reason">The reason to save the data.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="user"/> or <paramref name="item"/> or <paramref name="userData"/> or <paramref name="cancellationToken"/> is <c>null</c>.
+    /// </exception>
+    /// <returns>A <see cref="Task"/> to save the userdata against an item.</returns>
+    public async Task SaveUserDataAsync(User? user, BaseItem item, UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken)
+    {
+        ArgumentNullException.ThrowIfNull(userData);
+
+        ArgumentNullException.ThrowIfNull(item);
+
+        ArgumentNullException.ThrowIfNull(user);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var keys = item.GetUserDataKeys();
+
+        var dbContext = await _provider.CreateDbContextAsync(cancellationToken).ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            foreach (var key in keys)
+            {
+                userData.Key = key;
+                dbContext.UserDatas.Update(userData);
```

Comment:
I am not sure this does the trick. This should only mark the object as to be updated, not store the state. So the last value might win.